{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2970a8aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import sys, os\n",
    "from Bio import Entrez\n",
    "import json\n",
    "from ete3 import NCBITaxa\n",
    "import pandas as pd\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cae27eef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Declares\n",
    "CSV_FILE = \"../data/BDNF/BDNF_orthologs.csv\"\n",
    "#CSV_FILE = sys.argv[1]\n",
    "\n",
    "DATA_DICT = {}\n",
    "\n",
    "# The following looks up NCBI Taxon ID from Accession IDs\n",
    "# Then uses that Taxon ID to get the complete taxnomic lineage of of that species.\n",
    "Entrez.email = \"aglucaci@gmail.com\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9489bb68",
   "metadata": {},
   "outputs": [],
   "source": [
    "TREE_FILE = \"../results/BDNF/Recombinants/BDNF_codons_RDP_recombinationFree.fas.treefile\"\n",
    "\n",
    "#TREE_FILE = sys.argv[2]\n",
    "\n",
    "# Read file, get newick\n",
    "with open(TREE_FILE, \"r\") as fh:\n",
    "    TREE_NEWICK = fh.read()\n",
    "#end with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfcd3c4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'(NM_001709_5_Homo_sapiens_brain_derived_neurotrophi:0.0072488191,((((((((((((((NM_001270630_1_Rattus_norvegicus_brain_derived_neu:0.0002067937,(((NM_001048142_1_Mus_musculus_brain_derived_neurotro:0.0000025173,XM_021151236_2_PREDICTED_Mus_caroli_brain_derived_:0.0038244124):0.0029655172,XM_021192829_2_PREDICTED_Mus_pahari_brain_derived_:0.0146613311):0.0371460124,XM_032903861_1_PREDICTED_Rattus_rattus_brain_deriv:0.0042058297):0.0054785752):0.0777307590,(XM_007653166_4_PREDICTED_Cricetulus_griseus_brain_:0.0674809283,XM_037204438_1_PREDICTED_Peromyscus_leucopus_brain:0.0301779458):0.0262228836):0.0995228554,XM_029555090_1_PREDICTED_Nannospalax_galili_brain_:0.0659379589):0.0612090200,XM_004660846_2_PREDICTED_Jaculus_jaculus_brain_der:0.1959495753):0.0652542305,(((XM_007497196_2_PREDICTED_Monodelphis_domestica_bra:0.1208032517,((XM_020977976_1_PREDICTED_Phascolarctos_cinereus_br:0.1175236731,XM_027852578_1_PREDICTED_Vombatus_ursinus_brain_de:0.0381565277):0.0257429950,XM_036763631_1_PREDICTED_Trichosurus_vulpecula_bra:0.0618006448):0.0293292848):0.0123947508,XM_031944159_1_PREDICTED_Sarcophilus_harrisii_brai:0.1121180380):0.1282308172,(XM_029059317_2_PREDICTED_Ornithorhynchus_anatinus_:0.0534364749,XM_038764545_1_PREDICTED_Tachyglossus_aculeatus_br:0.0260205002):0.4096281680):0.2122617241):0.0370472488,(((((((((((XM_005654684_3_PREDICTED_Sus_scrofa_brain_derived_:0.0566589651,((((((XM_005216334_4_PREDICTED_Bos_taurus_brain_derived_:0.0032962461,(XM_010831788_1_PREDICTED_Bison_bison_bison_brain_d:0.0000010004,XM_005890667_2_PREDICTED_Bos_mutus_brain_derived_n:0.0000010004):0.0000026715):0.0000023389,XM_019974522_1_PREDICTED_Bos_indicus_brain_derived:0.0000010004):0.0035003548,XM_006068006_2_PREDICTED_Bubalus_bubalis_brain_der:0.0068546883):0.0032451195,XM_020903256_1_PREDICTED_Odocoileus_virginianus_te:0.0695475933):0.0033830418,((XM_012096129_4_PREDICTED_Ovis_aries_brain_derived_:0.0000022610,XM_005690025_2_PREDICTED_Capra_hircus_brain_derive:0.0099640895):0.0032731154,XM_040258919_1_PREDICTED_Oryx_dammah_brain_derived:0.0103035451):0.0037582468):0.0561480276,(((((((XM_024764694_1_PREDICTED_Neophocaena_asiaeoriental:0.0000026514,XM_032641481_1_PREDICTED_Phocoena_sinus_brain_deri:0.0071319078):0.0070276857,XM_029207934_1_PREDICTED_Monodon_monoceros_brain_d:0.0087015352):0.0000022358,XM_007461679_1_PREDICTED_Lipotes_vexillifer_brain_:0.0102799265):0.0000025691,XM_027115138_1_PREDICTED_Lagenorhynchus_obliquiden:0.0068299534):0.0000027323,XM_022583733_1_PREDICTED_Delphinapterus_leucas_bra:0.0081480470):0.0057669314,XM_028501134_1_PREDICTED_Physeter_catodon_brain_de:0.0146856048):0.0014556510,(XM_007180991_1_PREDICTED_Balaenoptera_acutorostrat:0.0098193882,XM_036862552_1_PREDICTED_Balaenoptera_musculus_bra:0.0009233804):0.0160294037):0.0640875633):0.0804380797):0.0218880593,((XM_031681074_1_PREDICTED_Vicugna_pacos_brain_deriv:0.0181915751,(XM_006195083_3_PREDICTED_Camelus_ferus_brain_deriv:0.0000021667,XM_031459858_1_PREDICTED_Camelus_dromedarius_brain:0.0040359604):0.0040391151):0.0000024424,XM_010964743_1_PREDICTED_Camelus_bactrianus_brain_:0.0000028352):0.0602276076):0.0287800463,((((((((XM_013147262_2_PREDICTED_Cavia_porcellus_brain_der:0.1046898533,(XM_005401751_2_PREDICTED_Chinchilla_lanigera_brain:0.0534896531,XM_004644113_2_PREDICTED_Octodon_degus_brain_deriv:0.0756464507):0.0398097797):0.0216026611,(XM_004851581_3_PREDICTED_Heterocephalus_glaber_bra:0.0729110157,XM_033764060_1_PREDICTED_Fukomys_damarensis_brain_:0.2230558244):0.0207124044):0.2302992947,XM_036427676_1_PREDICTED_Pipistrellus_kuhlii_brain:0.1248434560):0.1182435294,XM_008148947_2_PREDICTED_Eptesicus_fuscus_brain_de:0.0000027558):0.0516072910,XM_036325692_1_PREDICTED_Myotis_myotis_brain_deriv:0.0063520424):0.0109878687,XM_014545223_1_PREDICTED_Myotis_brandtii_brain_der:0.0000027115):0.0029149851,(XM_006753948_2_PREDICTED_Myotis_davidii_brain_deri:0.0248505781,XM_023756374_1_PREDICTED_Myotis_lucifugus_brain_de:0.0034609720):0.0000026133):0.0616951086,(XM_016206154_1_PREDICTED_Miniopterus_natalensis_br:0.2625017777,XM_036258707_1_PREDICTED_Molossus_molossus_brain_d:0.0164164348):0.0127613754):0.0145909590):0.0020983135,(XM_036879079_1_PREDICTED_Manis_pentadactyla_brain_:0.0160739871,XM_017667612_2_PREDICTED_Manis_javanica_brain_deri:0.0090292159):0.1091591614):0.0070580080,((NM_001081787_1_Equus_caballus_brain_derived_neurot:0.0000010004,XM_014831462_1_PREDICTED_Equus_asinus_brain_derive:0.0035756048):0.0619062678,XM_004418542_2_PREDICTED_Ceratotherium_simum_simum:0.0316268039):0.0853496163):0.0007884031,(XM_004682964_2_PREDICTED_Condylura_cristata_brain_:0.0575781497,XM_037522538_1_PREDICTED_Talpa_occidentalis_brain_:0.0287314209):0.0571328241):0.0032633348,((((((NM_001009828_1_Felis_catus_brain_derived_neurotrop:0.0182785660,XM_030333208_1_PREDICTED_Lynx_canadensis_brain_der:0.0037147112):0.0056645059,XM_015544869_1_PREDICTED_Panthera_tigris_altaica_b:0.0085160848):0.0187844633,XM_029957152_1_PREDICTED_Suricata_suricatta_brain_:0.0221791056):0.0037424374,XM_039217098_1_PREDICTED_Hyaena_hyaena_brain_deriv:0.0189732462):0.0238256407,XM_016194241_1_PREDICTED_Erinaceus_europaeus_brain:0.1852411177):0.0024489746,(((XM_038429434_1_PREDICTED_Canis_lupus_familiaris_br:0.0000020785,(XM_041730708_1_PREDICTED_Vulpes_lagopus_brain_deri:0.0000010004,XM_026018600_1_PREDICTED_Vulpes_vulpes_brain_deriv:0.0000010004):0.0034531265):0.0864845386,((XM_011226480_3_PREDICTED_Ailuropoda_melanoleuca_br:0.0045871410,((XM_004408653_1_PREDICTED_Odobenus_rosmarus_diverge:0.0034081915,(XM_025879326_1_PREDICTED_Callorhinus_ursinus_brain:0.0071345814,(XM_028117708_1_PREDICTED_Eumetopias_jubatus_brain_:0.0000010004,XM_027577807_1_PREDICTED_Zalophus_californianus_br:0.0034362819):0.0034617350):0.0034994095):0.0105092387,((XM_021685880_1_PREDICTED_Neomonachus_schauinslandi:0.0000022958,XM_031025064_1_PREDICTED_Leptonychotes_weddellii_b:0.0034481263):0.0034436703,((XM_036109493_1_PREDICTED_Halichoerus_grypus_brain_:0.0000024259,XM_032414183_1_PREDICTED_Phoca_vitulina_brain_deri:0.0056758082):0.0054578148,XM_035012477_1_PREDICTED_Mirounga_leonina_brain_de:0.0180572880):0.0000027434):0.0114569600):0.0429628239):0.0029096528,XM_026512203_1_PREDICTED_Ursus_arctos_horribilis_b:0.0113503994):0.0223499921):0.0099530151,((((XM_011362565_2_PREDICTED_Pteropus_vampyrus_brain_d:0.0000022809,XM_039866065_1_PREDICTED_Pteropus_giganteus_brain_:0.0035758184):0.0036430019,XM_006907988_2_PREDICTED_Pteropus_alecto_brain_der:0.0033058156):0.0402634751,(((XM_037040421_1_PREDICTED_Sturnira_hondurensis_brai:0.0518556927,XM_037146035_1_PREDICTED_Artibeus_jamaicensis_brai:0.0325434407):0.0064780078,XM_024558903_1_PREDICTED_Desmodus_rotundus_brain_d:0.0160318139):0.0000028912,XM_028517638_2_PREDICTED_Phyllostomus_discolor_bra:0.0093682224):0.0593618986):0.0013612485,(XM_033118944_1_PREDICTED_Rhinolophus_ferrumequinum:0.0634471598,XM_019650718_1_PREDICTED_Hipposideros_armiger_brai:0.0704703594):0.0466994927):0.0391312145):0.0078599373):0.0159287681):0.0084089420,((XM_012649556_1_PREDICTED_Propithecus_coquereli_bra:0.0190289711,XM_020286534_1_PREDICTED_Microcebus_murinus_brain_:0.0090030018):0.0081770017,XM_012814047_1_PREDICTED_Otolemur_garnettii_brain_:0.0393970501):0.0237002895):0.0041233660,(((XM_009007854_3_PREDICTED_Callithrix_jacchus_brain_:0.0303066703,(XM_015476449_1_PREDICTED_Marmota_marmota_marmota_b:0.0000027824,((XM_040284582_1_PREDICTED_Ictidomys_tridecemlineatu:0.0141929763,XM_026412272_1_PREDICTED_Urocitellus_parryii_brain:0.0133774604):0.0291095485,XM_027936596_1_PREDICTED_Marmota_flaviventris_brai:0.0000020471):0.0092331879):0.0320946293):0.0422043083,(XM_013019592_1_PREDICTED_Dipodomys_ordii_brain_der:0.0536587912,XM_020177315_1_PREDICTED_Castor_canadensis_brain_d:0.0324951919):0.0380105017):0.0000028862,XM_031371313_1_PREDICTED_Mastomys_coucha_brain_der:0.0562317007):0.0110698552):0.0267872803,XM_008050727_1_PREDICTED_Carlito_syrichta_brain_de:0.1655516332):0.0577225812,XM_014591303_2_PREDICTED_Tupaia_chinensis_brain_de:0.1486640507):0.0408624351):0.0543648839,(XM_007951952_1_PREDICTED_Orycteropus_afer_afer_bra:0.0699922753,XM_004369733_1_PREDICTED_Trichechus_manatus_latiro:0.0184391736):0.0031687492):0.0405913675,(XM_006883713_1_PREDICTED_Elephantulus_edwardii_bra:0.0475243894,(XM_004708253_2_PREDICTED_Echinops_telfairi_brain_d:0.1614297408,XM_006864851_1_PREDICTED_Chrysochloris_asiatica_br:0.0499603033):0.0115833377):0.0056684364):0.0448772656,(XM_037839114_1_PREDICTED_Choloepus_didactylus_brai:0.0326537520,XM_012523604_2_PREDICTED_Dasypus_novemcinctus_brai:0.0568520452):0.0295591697):0.0031565047,((XM_040980297_1_PREDICTED_Ochotona_curzoniae_brain_:0.0139796787,XM_004585440_1_PREDICTED_Ochotona_princeps_brain_d:0.0000027693):0.1427492110,XM_008584165_1_PREDICTED_Galeopterus_variegatus_br:0.0179243581):0.0138730916):0.0384773590,((XM_015114598_2_PREDICTED_Macaca_mulatta_brain_deri:0.0013664488,(((XM_011936307_1_PREDICTED_Colobus_angolensis_pallia:0.0000022514,XM_023230347_2_PREDICTED_Piliocolobus_tephrosceles:0.0000010004):0.0013649473,(XM_033205475_1_PREDICTED_Trachypithecus_francoisi_:0.0027378466,(XM_017858805_1_PREDICTED_Rhinopithecus_bieti_brain:0.0013648598,XM_010374529_2_PREDICTED_Rhinopithecus_roxellana_b:0.0000026646):0.0027353246):0.0000023184):0.0027429401,(XM_008003703_2_PREDICTED_Chlorocebus_sabaeus_brain:0.0041169339,XM_017948537_3_PREDICTED_Papio_anubis_brain_derive:0.0027385592):0.0000024829):0.0000025307):0.0013734781,(XM_012031697_1_PREDICTED_Cercocebus_atys_brain_der:0.0013763298,XM_011964715_1_PREDICTED_Mandrillus_leucophaeus_br:0.0027485757):0.0013631712):0.0019131872):0.0055537271,((XM_003919940_3_PREDICTED_Saimiri_boliviensis_boliv:0.0214786868,(XM_032283691_1_PREDICTED_Sapajus_apella_brain_deri:0.0027197837,XM_017538722_1_PREDICTED_Cebus_imitator_brain_deri:0.0013736736):0.0017085739):0.0021185128,XM_012452238_1_PREDICTED_Aotus_nancymaae_brain_der:0.0091343756):0.0049733747):0.0032851523,(XM_002821931_2_PREDICTED_Pongo_abelii_brain_derive:0.0097838505,(XM_003254347_2_PREDICTED_Nomascus_leucogenys_brain:0.0013751904,XM_032166370_1_PREDICTED_Hylobates_moloch_brain_de:0.0013696226):0.0027625468):0.0014947870):0.0661607115,XM_004050851_2_PREDICTED_Gorilla_gorilla_gorilla_b:0.0035734281):0.0000023922,(NM_001012441_1_Pan_troglodytes_brain_derived_neuro:0.0035734414,XM_034932318_1_PREDICTED_Pan_paniscus_brain_derive:0.0172746495):0.0174993818);\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TREE_NEWICK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae947ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific name</th>\n",
       "      <th>Common name</th>\n",
       "      <th>RefSeq Transcript accessions</th>\n",
       "      <th>RefSeq Protein accessions</th>\n",
       "      <th>SPARCLE ID</th>\n",
       "      <th>Curation Status</th>\n",
       "      <th>RefseqSelect</th>\n",
       "      <th>Default sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>627</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NM_001709.5</td>\n",
       "      <td>NP_001700.2</td>\n",
       "      <td>10639753.0</td>\n",
       "      <td>curated</td>\n",
       "      <td>RefseqSelect</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12064</td>\n",
       "      <td>Bdnf</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>NM_001048142.1</td>\n",
       "      <td>NP_001041607.1</td>\n",
       "      <td>10639753.0</td>\n",
       "      <td>curated</td>\n",
       "      <td>RefseqSelect</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24225</td>\n",
       "      <td>Bdnf</td>\n",
       "      <td>brain-derived neurotrophic factor</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Norway rat</td>\n",
       "      <td>NM_001270630.1</td>\n",
       "      <td>NP_001257559.1</td>\n",
       "      <td>10639753.0</td>\n",
       "      <td>curated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>397495</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Sus scrofa</td>\n",
       "      <td>pig</td>\n",
       "      <td>XM_005654684.3</td>\n",
       "      <td>XP_005654741.1</td>\n",
       "      <td>10639753.0</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>403461</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>dog</td>\n",
       "      <td>XM_038429434.1</td>\n",
       "      <td>XP_038285362.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>120876831</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Oryx dammah</td>\n",
       "      <td>scimitar-horned oryx</td>\n",
       "      <td>XM_040258919.1</td>\n",
       "      <td>XP_040114853.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>121043625</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Puma yagouaroundi</td>\n",
       "      <td>jaguarundi</td>\n",
       "      <td>XM_040495572.1</td>\n",
       "      <td>XP_040351506.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>121156520</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Ochotona curzoniae</td>\n",
       "      <td>black-lipped pika</td>\n",
       "      <td>XM_040980297.1</td>\n",
       "      <td>XP_040836231.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>121465693</td>\n",
       "      <td>Bdnf</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Microtus oregoni</td>\n",
       "      <td>creeping vole</td>\n",
       "      <td>XM_041679035.1</td>\n",
       "      <td>XP_041534969.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>121476611</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>brain derived neurotrophic factor</td>\n",
       "      <td>Vulpes lagopus</td>\n",
       "      <td>Arctic fox</td>\n",
       "      <td>XM_041730708.1</td>\n",
       "      <td>XP_041586642.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>model</td>\n",
       "      <td>NaN</td>\n",
       "      <td>default-seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene ID Gene symbol                        Description  \\\n",
       "1          627        BDNF  brain derived neurotrophic factor   \n",
       "2        12064        Bdnf  brain derived neurotrophic factor   \n",
       "3        24225        Bdnf  brain-derived neurotrophic factor   \n",
       "4       397495        BDNF  brain derived neurotrophic factor   \n",
       "5       403461        BDNF  brain derived neurotrophic factor   \n",
       "..         ...         ...                                ...   \n",
       "158  120876831        BDNF  brain derived neurotrophic factor   \n",
       "159  121043625        BDNF  brain derived neurotrophic factor   \n",
       "160  121156520        BDNF  brain derived neurotrophic factor   \n",
       "161  121465693        Bdnf  brain derived neurotrophic factor   \n",
       "162  121476611        BDNF  brain derived neurotrophic factor   \n",
       "\n",
       "            Scientific name           Common name  \\\n",
       "1              Homo sapiens                 human   \n",
       "2              Mus musculus           house mouse   \n",
       "3         Rattus norvegicus            Norway rat   \n",
       "4                Sus scrofa                   pig   \n",
       "5    Canis lupus familiaris                   dog   \n",
       "..                      ...                   ...   \n",
       "158             Oryx dammah  scimitar-horned oryx   \n",
       "159       Puma yagouaroundi            jaguarundi   \n",
       "160      Ochotona curzoniae     black-lipped pika   \n",
       "161        Microtus oregoni         creeping vole   \n",
       "162          Vulpes lagopus            Arctic fox   \n",
       "\n",
       "    RefSeq Transcript accessions RefSeq Protein accessions  SPARCLE ID  \\\n",
       "1                    NM_001709.5               NP_001700.2  10639753.0   \n",
       "2                 NM_001048142.1            NP_001041607.1  10639753.0   \n",
       "3                 NM_001270630.1            NP_001257559.1  10639753.0   \n",
       "4                 XM_005654684.3            XP_005654741.1  10639753.0   \n",
       "5                 XM_038429434.1            XP_038285362.1         NaN   \n",
       "..                           ...                       ...         ...   \n",
       "158               XM_040258919.1            XP_040114853.1         NaN   \n",
       "159               XM_040495572.1            XP_040351506.1         NaN   \n",
       "160               XM_040980297.1            XP_040836231.1         NaN   \n",
       "161               XM_041679035.1            XP_041534969.1         NaN   \n",
       "162               XM_041730708.1            XP_041586642.1         NaN   \n",
       "\n",
       "    Curation Status  RefseqSelect Default sequence  \n",
       "1           curated  RefseqSelect      default-seq  \n",
       "2           curated  RefseqSelect      default-seq  \n",
       "3           curated           NaN      default-seq  \n",
       "4             model           NaN      default-seq  \n",
       "5             model           NaN      default-seq  \n",
       "..              ...           ...              ...  \n",
       "158           model           NaN      default-seq  \n",
       "159           model           NaN      default-seq  \n",
       "160           model           NaN      default-seq  \n",
       "161           model           NaN      default-seq  \n",
       "162           model           NaN      default-seq  \n",
       "\n",
       "[162 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(CSV_FILE)\n",
    "df.index += 1\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7209052",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_accessions = df['RefSeq Transcript accessions'].tolist()\n",
    "\n",
    "#len(transcript_accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c521870",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function\n",
    "def match_transcript_to_tree(TREE_NEWICK, accession):\n",
    "    t = Tree(TREE_NEWICK, format=1)\n",
    "    \n",
    "    # uncomment to debug\n",
    "    #print(\"# in 'match_transcript_to_tree'\", accession)\n",
    "    \n",
    "    for leafname in t.get_leaf_names():\n",
    "        if accession in leafname:\n",
    "            # uncomment to debug\n",
    "            #print(\"# Returning:\", leafname)\n",
    "            return leafname\n",
    "        #end if\n",
    "    #end for\n",
    "#end match\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a29a8695",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(transcript_accessions):\n",
    "    #DATA_DICT = {}\n",
    "    global DATA_DICT, TREE_NEWICK\n",
    "    count = 1\n",
    "    for ACCESSION in transcript_accessions:\n",
    "        #print(\"# Checking ACCESION:\", ACCESSION)\n",
    "        skip = False\n",
    "        for i in DATA_DICT.keys():\n",
    "            if ACCESSION == DATA_DICT[i][\"ACCESSION\"]:\n",
    "                skip = True\n",
    "                break\n",
    "            #end if\n",
    "        #end for\n",
    "        if skip == True: \n",
    "            count += 1\n",
    "            continue\n",
    "        #end if\n",
    "        \n",
    "        try:\n",
    "            handle = Entrez.esummary(db=\"nucleotide\", id=ACCESSION, rettype=\"gb\", retmode=\"text\", retmax=1)\n",
    "            records = Entrez.parse(handle)\n",
    "        except Exception as e:\n",
    "            print(\"# Error, sleeping\", e)\n",
    "            time.sleep(5)\n",
    "            handle = Entrez.esummary(db=\"nucleotide\", id=ACCESSION, rettype=\"gb\", retmode=\"text\", retmax=1)\n",
    "            records = Entrez.parse(handle)\n",
    "        #end try\n",
    "        \n",
    "        try:\n",
    "            for record in records:\n",
    "                TAXON_ID = record[\"TaxId\"]\n",
    "                \n",
    "                print(\"#\", count, \"Processing transcript accession:\", str(ACCESSION), \"with NCBI Taxon ID:\", str(TAXON_ID))\n",
    "                ncbi = NCBITaxa()\n",
    "                \n",
    "                lineage = ncbi.get_lineage(TAXON_ID)\n",
    "                # uncomment to debug\n",
    "                #print(lineage) # returns a list of taxon id's\n",
    "                \n",
    "                names = ncbi.get_taxid_translator(lineage)\n",
    "                # uncomment to debug\n",
    "                #print(names) # translates the taxon id's into clade names\n",
    "                \n",
    "                # uncomment to debug\n",
    "                #print(\"# ACCESSION:\", ACCESSION)\n",
    "                leafname = \"\"\n",
    "                leafname = match_transcript_to_tree(TREE_NEWICK, ACCESSION.replace(\".\", \"_\"))\n",
    "                \n",
    "                # uncomment to debug\n",
    "                #print(leafname)\n",
    "                \n",
    "                DATA_DICT[str(count)] = {\"ACCESSION\": ACCESSION, \"TAXON_ID\": TAXON_ID, \n",
    "                                         \"LINEAGE\": [names[taxid] for taxid in lineage], \n",
    "                                         \"TITLE\":record[\"Title\"], \"LEAFNAME\": leafname}\n",
    "                count += 1\n",
    "            #end inner for\n",
    "            handle.close\n",
    "        except Exception as e:\n",
    "            print(\"# Error (main):\", ACCESSION, e, \"\\n\")\n",
    "            #print(names)\n",
    "            #print(e)\n",
    "            #main(TREE_NEWICK) # Recursive.\n",
    "            #break\n",
    "        #end try\n",
    "        \n",
    "    #end outer for\n",
    "    \n",
    "    #return DATA_DICT\n",
    "#end method\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f18ccfec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# 1 Processing transcript accession: NM_001709.5 with NCBI Taxon ID: 9606\n",
      "# 2 Processing transcript accession: NM_001048142.1 with NCBI Taxon ID: 10090\n",
      "# 3 Processing transcript accession: NM_001270630.1 with NCBI Taxon ID: 10116\n",
      "# 4 Processing transcript accession: XM_005654684.3 with NCBI Taxon ID: 9823\n",
      "# 5 Processing transcript accession: XM_038429434.1 with NCBI Taxon ID: 9615\n",
      "# 6 Processing transcript accession: NM_001009828.1 with NCBI Taxon ID: 9685\n",
      "# 7 Processing transcript accession: NM_001012441.1 with NCBI Taxon ID: 9598\n",
      "# 8 Processing transcript accession: XM_007497196.2 with NCBI Taxon ID: 13616\n",
      "# 9 Processing transcript accession: XM_005216334.4 with NCBI Taxon ID: 9913\n",
      "# 10 Processing transcript accession: XM_015114598.2 with NCBI Taxon ID: 9544\n",
      "# 11 Processing transcript accession: NM_001081787.1 with NCBI Taxon ID: 9796\n",
      "# 12 Processing transcript accession: XM_029059317.2 with NCBI Taxon ID: 9258\n",
      "# 13 Processing transcript accession: XM_017345633.1 with NCBI Taxon ID: 9986\n",
      "# 14 Processing transcript accession: XM_009007854.3 with NCBI Taxon ID: 9483\n",
      "# 15 Processing transcript accession: XM_002821931.2 with NCBI Taxon ID: 9601\n",
      "# 16 Processing transcript accession: XM_011226480.3 with NCBI Taxon ID: 9646\n",
      "# 17 Processing transcript accession: XM_003254347.2 with NCBI Taxon ID: 61853\n",
      "# 18 Processing transcript accession: XM_023550772.1 with NCBI Taxon ID: 9785\n",
      "# 19 Processing transcript accession: XM_013147262.2 with NCBI Taxon ID: 10141\n",
      "# 20 Processing transcript accession: XM_007653166.4 with NCBI Taxon ID: 10029\n",
      "# 21 Processing transcript accession: XM_031944159.1 with NCBI Taxon ID: 9305\n",
      "# 22 Processing transcript accession: XM_012814047.1 with NCBI Taxon ID: 30611\n",
      "# 23 Processing transcript accession: XM_034932318.1 with NCBI Taxon ID: 9597\n",
      "# 24 Processing transcript accession: XM_017948537.3 with NCBI Taxon ID: 9555\n",
      "# 25 Processing transcript accession: XM_003919940.3 with NCBI Taxon ID: 39432\n",
      "# 26 Processing transcript accession: XM_012096129.4 with NCBI Taxon ID: 9940\n",
      "# 27 Processing transcript accession: XM_004050851.2 with NCBI Taxon ID: 9595\n",
      "# 28 Processing transcript accession: XM_004263941.3 with NCBI Taxon ID: 9733\n",
      "# 29 Processing transcript accession: XM_019947883.2 with NCBI Taxon ID: 9739\n",
      "# 30 Processing transcript accession: XM_004369733.1 with NCBI Taxon ID: 127582\n",
      "# 31 Processing transcript accession: XM_004408653.1 with NCBI Taxon ID: 9708\n",
      "# 32 Processing transcript accession: XM_004418542.2 with NCBI Taxon ID: 73337\n",
      "# 33 Processing transcript accession: XM_012523604.2 with NCBI Taxon ID: 9361\n",
      "# 34 Processing transcript accession: XM_004585440.1 with NCBI Taxon ID: 9978\n",
      "# 35 Processing transcript accession: XM_004607860.1 with NCBI Taxon ID: 42254\n",
      "# 36 Processing transcript accession: XM_004644113.2 with NCBI Taxon ID: 10160\n",
      "# 37 Processing transcript accession: XM_004660846.2 with NCBI Taxon ID: 51337\n",
      "# 38 Processing transcript accession: XM_004682964.2 with NCBI Taxon ID: 143302\n",
      "# 39 Processing transcript accession: XM_004708253.2 with NCBI Taxon ID: 9371\n",
      "# 40 Processing transcript accession: XM_004755891.2 with NCBI Taxon ID: 9669\n",
      "# 41 Processing transcript accession: XM_004851581.3 with NCBI Taxon ID: 10181\n",
      "# 42 Processing transcript accession: XM_005064810.4 with NCBI Taxon ID: 10036\n",
      "# 43 Processing transcript accession: XM_040284582.1 with NCBI Taxon ID: 43179\n",
      "# 44 Processing transcript accession: XM_005364052.2 with NCBI Taxon ID: 79684\n",
      "# 45 Processing transcript accession: XM_005401751.2 with NCBI Taxon ID: 34839\n",
      "# 46 Processing transcript accession: XM_005578346.2 with NCBI Taxon ID: 9541\n",
      "# 47 Processing transcript accession: XM_005690025.2 with NCBI Taxon ID: 9925\n",
      "# 48 Processing transcript accession: XM_014545223.1 with NCBI Taxon ID: 109478\n",
      "# 49 Processing transcript accession: XM_005890667.2 with NCBI Taxon ID: 72004\n",
      "# 50 Processing transcript accession: XM_006068006.2 with NCBI Taxon ID: 89462\n",
      "# 51 Processing transcript accession: XM_023756374.1 with NCBI Taxon ID: 59463\n",
      "# 52 Processing transcript accession: XM_014591303.2 with NCBI Taxon ID: 246437\n",
      "# 53 Processing transcript accession: XM_006195083.3 with NCBI Taxon ID: 419612\n",
      "# 54 Processing transcript accession: XM_031681074.1 with NCBI Taxon ID: 30538\n",
      "# 55 Processing transcript accession: XM_031025064.1 with NCBI Taxon ID: 9713\n",
      "# 56 Processing transcript accession: XM_006753948.2 with NCBI Taxon ID: 225400\n",
      "# 57 Processing transcript accession: XM_006864851.1 with NCBI Taxon ID: 185453\n",
      "# 58 Processing transcript accession: XM_006883713.1 with NCBI Taxon ID: 28737\n",
      "# 59 Processing transcript accession: XM_006907988.2 with NCBI Taxon ID: 9402\n",
      "# 60 Processing transcript accession: XM_006980818.2 with NCBI Taxon ID: 230844\n",
      "# 61 Processing transcript accession: XM_015544869.1 with NCBI Taxon ID: 74533\n",
      "# 62 Processing transcript accession: XM_028501134.1 with NCBI Taxon ID: 9755\n",
      "# 63 Processing transcript accession: XM_007180991.1 with NCBI Taxon ID: 310752\n",
      "# 64 Processing transcript accession: XM_007461679.1 with NCBI Taxon ID: 118797\n",
      "# 65 Processing transcript accession: XM_016194241.1 with NCBI Taxon ID: 9365\n",
      "# 66 Processing transcript accession: XM_007951952.1 with NCBI Taxon ID: 1230840\n",
      "# 67 Processing transcript accession: XM_008003703.2 with NCBI Taxon ID: 60711\n",
      "# 68 Processing transcript accession: XM_008050727.1 with NCBI Taxon ID: 1868482\n",
      "# 69 Processing transcript accession: XM_008148947.2 with NCBI Taxon ID: 29078\n",
      "# 70 Processing transcript accession: XM_008522855.1 with NCBI Taxon ID: 9798\n",
      "# 71 Processing transcript accession: XM_008584165.1 with NCBI Taxon ID: 482537\n",
      "# 72 Processing transcript accession: XM_008686880.2 with NCBI Taxon ID: 29073\n",
      "# 73 Processing transcript accession: XM_029555090.1 with NCBI Taxon ID: 1026970\n",
      "# 74 Processing transcript accession: XM_010374529.2 with NCBI Taxon ID: 61622\n",
      "# 75 Processing transcript accession: XM_033764060.1 with NCBI Taxon ID: 885580\n",
      "# 76 Processing transcript accession: XM_010831788.1 with NCBI Taxon ID: 43346\n",
      "# 77 Processing transcript accession: XM_010964743.1 with NCBI Taxon ID: 9837\n",
      "# 78 Processing transcript accession: XM_031459858.1 with NCBI Taxon ID: 9838\n",
      "# 79 Processing transcript accession: XM_011362565.2 with NCBI Taxon ID: 132908\n",
      "# 80 Processing transcript accession: XM_011710310.2 with NCBI Taxon ID: 9545\n",
      "# 81 Processing transcript accession: XM_011936307.1 with NCBI Taxon ID: 336983\n",
      "# 82 Processing transcript accession: XM_011964715.1 with NCBI Taxon ID: 9568\n",
      "# 83 Processing transcript accession: XM_012031697.1 with NCBI Taxon ID: 9531\n",
      "# 84 Processing transcript accession: XM_012452238.1 with NCBI Taxon ID: 37293\n",
      "# 85 Processing transcript accession: XM_012649556.1 with NCBI Taxon ID: 379532\n",
      "# 86 Processing transcript accession: XM_020286534.1 with NCBI Taxon ID: 30608\n",
      "# 87 Processing transcript accession: XM_013019592.1 with NCBI Taxon ID: 10020\n",
      "# 88 Processing transcript accession: XM_014831462.1 with NCBI Taxon ID: 9793\n",
      "# 89 Processing transcript accession: XM_027072960.1 with NCBI Taxon ID: 32536\n",
      "# 90 Processing transcript accession: XM_015476449.1 with NCBI Taxon ID: 9994\n",
      "# 91 Processing transcript accession: XM_036221046.1 with NCBI Taxon ID: 9407\n",
      "# 92 Processing transcript accession: XM_016206154.1 with NCBI Taxon ID: 291302\n",
      "# 93 Processing transcript accession: XM_017538722.1 with NCBI Taxon ID: 2715852\n",
      "# Error (main): XM_017538722.1 2715852 taxid not found \n",
      "\n",
      "# 93 Processing transcript accession: XM_017667612.2 with NCBI Taxon ID: 9974\n",
      "# 94 Processing transcript accession: XM_017858805.1 with NCBI Taxon ID: 61621\n",
      "# 95 Processing transcript accession: XM_019456577.1 with NCBI Taxon ID: 9691\n",
      "# 96 Processing transcript accession: XM_019650718.1 with NCBI Taxon ID: 186990\n",
      "# 97 Processing transcript accession: XM_019974522.1 with NCBI Taxon ID: 9915\n",
      "# 98 Processing transcript accession: XM_020177315.1 with NCBI Taxon ID: 51338\n",
      "# 99 Processing transcript accession: XM_020903256.1 with NCBI Taxon ID: 9880\n",
      "# 100 Processing transcript accession: XM_020977976.1 with NCBI Taxon ID: 38626\n",
      "# 101 Processing transcript accession: XM_021151236.2 with NCBI Taxon ID: 10089\n",
      "# 102 Processing transcript accession: XM_021192829.2 with NCBI Taxon ID: 10093\n",
      "# 103 Processing transcript accession: XM_021685880.1 with NCBI Taxon ID: 29088\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# 104 Processing transcript accession: XM_022507574.1 with NCBI Taxon ID: 391180\n",
      "# 105 Processing transcript accession: XM_022583733.1 with NCBI Taxon ID: 9749\n",
      "# 106 Processing transcript accession: XM_023230347.2 with NCBI Taxon ID: 591936\n",
      "# 107 Processing transcript accession: XM_024558903.1 with NCBI Taxon ID: 9430\n",
      "# 108 Processing transcript accession: XM_024764694.1 with NCBI Taxon ID: 1706337\n",
      "# 109 Processing transcript accession: XM_025358305.1 with NCBI Taxon ID: 9565\n",
      "# 110 Processing transcript accession: XM_025459816.2 with NCBI Taxon ID: 286419\n",
      "# 111 Processing transcript accession: XM_025879326.1 with NCBI Taxon ID: 34884\n",
      "# 112 Processing transcript accession: XM_025929035.1 with NCBI Taxon ID: 9696\n",
      "# 113 Processing transcript accession: XM_026018600.1 with NCBI Taxon ID: 9627\n",
      "# 114 Processing transcript accession: XM_026412272.1 with NCBI Taxon ID: 9999\n",
      "# 115 Processing transcript accession: XM_026512203.1 with NCBI Taxon ID: 116960\n",
      "# 116 Processing transcript accession: XM_027115138.1 with NCBI Taxon ID: 90247\n",
      "# 117 Processing transcript accession: XM_027562786.1 with NCBI Taxon ID: 30522\n",
      "# 118 Processing transcript accession: XM_027577807.1 with NCBI Taxon ID: 9704\n",
      "# 119 Processing transcript accession: XM_027852578.1 with NCBI Taxon ID: 29139\n",
      "# 120 Processing transcript accession: XM_027936596.1 with NCBI Taxon ID: 93162\n",
      "# 121 Processing transcript accession: XM_028117708.1 with NCBI Taxon ID: 34886\n",
      "# 122 Processing transcript accession: XM_028517638.2 with NCBI Taxon ID: 89673\n",
      "# 123 Processing transcript accession: XM_028766921.1 with NCBI Taxon ID: 491861\n",
      "# 124 Processing transcript accession: XM_037204438.1 with NCBI Taxon ID: 10041\n",
      "# 125 Processing transcript accession: XM_029207934.1 with NCBI Taxon ID: 40151\n",
      "# 126 Processing transcript accession: XM_029957152.1 with NCBI Taxon ID: 37032\n",
      "# 127 Processing transcript accession: XM_030333208.1 with NCBI Taxon ID: 61383\n",
      "# 128 Processing transcript accession: XM_030864833.1 with NCBI Taxon ID: 9731\n",
      "# 129 Processing transcript accession: XM_031371313.1 with NCBI Taxon ID: 35658\n",
      "# 130 Processing transcript accession: XM_032166370.1 with NCBI Taxon ID: 81572\n",
      "# 131 Processing transcript accession: XM_032283691.1 with NCBI Taxon ID: 9515\n",
      "# 132 Processing transcript accession: XM_032358760.1 with NCBI Taxon ID: 36723\n",
      "# 133 Processing transcript accession: XM_032414183.1 with NCBI Taxon ID: 9720\n",
      "# 134 Processing transcript accession: XM_032641481.1 with NCBI Taxon ID: 42100\n",
      "# 135 Processing transcript accession: XM_032881246.1 with NCBI Taxon ID: 76717\n",
      "# 136 Processing transcript accession: XM_032903861.1 with NCBI Taxon ID: 10117\n",
      "# 137 Processing transcript accession: XM_033118944.1 with NCBI Taxon ID: 59479\n",
      "# 138 Processing transcript accession: XM_033205475.1 with NCBI Taxon ID: 54180\n",
      "# 139 Processing transcript accession: XM_034496309.1 with NCBI Taxon ID: 61156\n",
      "# 140 Processing transcript accession: XM_035012477.1 with NCBI Taxon ID: 9715\n",
      "# 141 Processing transcript accession: XM_036109493.1 with NCBI Taxon ID: 9711\n",
      "# 142 Processing transcript accession: XM_036184983.1 with NCBI Taxon ID: 38674\n",
      "# 143 Processing transcript accession: XM_036258707.1 with NCBI Taxon ID: 27622\n",
      "# 144 Processing transcript accession: XM_036325692.1 with NCBI Taxon ID: 51298\n",
      "# 145 Processing transcript accession: XM_036427676.1 with NCBI Taxon ID: 59472\n",
      "# 146 Processing transcript accession: XM_036763631.1 with NCBI Taxon ID: 9337\n",
      "# 147 Processing transcript accession: XM_036862552.1 with NCBI Taxon ID: 9771\n",
      "# 148 Processing transcript accession: XM_036879079.1 with NCBI Taxon ID: 143292\n",
      "# 149 Processing transcript accession: XM_037040421.1 with NCBI Taxon ID: 192404\n",
      "# 150 Processing transcript accession: XM_037146035.1 with NCBI Taxon ID: 9417\n",
      "# 151 Processing transcript accession: XM_037522538.1 with NCBI Taxon ID: 50954\n",
      "# 152 Processing transcript accession: XM_037839114.1 with NCBI Taxon ID: 27675\n",
      "# 153 Processing transcript accession: XM_038331579.1 with NCBI Taxon ID: 1047088\n",
      "# 154 Processing transcript accession: XM_038764545.1 with NCBI Taxon ID: 9261\n",
      "# 155 Processing transcript accession: XM_039217098.1 with NCBI Taxon ID: 95912\n",
      "# 156 Processing transcript accession: XM_039866065.1 with NCBI Taxon ID: 143291\n",
      "# 157 Processing transcript accession: XM_040258919.1 with NCBI Taxon ID: 59534\n",
      "# 158 Processing transcript accession: XM_040495572.1 with NCBI Taxon ID: 1608482\n",
      "# 159 Processing transcript accession: XM_040980297.1 with NCBI Taxon ID: 130825\n",
      "# 160 Processing transcript accession: XM_041679035.1 with NCBI Taxon ID: 111838\n",
      "# 161 Processing transcript accession: XM_041730708.1 with NCBI Taxon ID: 494514\n"
     ]
    }
   ],
   "source": [
    "main(transcript_accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "80b87b48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "162\n",
      "161\n"
     ]
    }
   ],
   "source": [
    "# Check for errors here\n",
    "print(len(transcript_accessions)) # Total\n",
    "print(len(DATA_DICT.keys())) # What we have identified\n",
    "# They need to match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "47137309",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.DataFrame.from_dict(DATA_DICT, orient=\"index\")\n",
    "#df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "55862dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df2[\"LINEAGE\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e203fde7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='LINEAGE'>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use explode to expand the lists into separate rows\n",
    "dfe = df2.LINEAGE.explode().to_frame().reset_index(drop=True)\n",
    "\n",
    "## groupby the values in the column, get the count and sort\n",
    "#dfg = dfe.groupby('LINEAGE').count() \\\n",
    "#                               .reset_index(name='count') \\\n",
    "#                               .sort_values(['count'], ascending=False) #\n",
    "\n",
    "#display(dfg)\n",
    "#dfg = dfe.groupby('LINEAGE').count().reset_index().sort_values(['count'], ascending=False) \n",
    "#dfg \n",
    "\n",
    "# plot the dataframe\n",
    "#dfg.plot.bar(x='LINEAGE')\n",
    "\n",
    "dfe\n",
    "\n",
    "dfg = dfe.groupby('LINEAGE').LINEAGE.count() \\\n",
    "                               .reset_index(name='count') \\\n",
    "                               .sort_values(['count'], ascending=False) \\\n",
    "                               .head(50).reset_index(drop=True)\n",
    "\n",
    "dfg.plot.barh(x='LINEAGE', figsize=(12, 12))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "82303212",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 1\n",
      "1 1\n",
      "2 1\n",
      "3 1\n",
      "4 1\n",
      "5 1\n",
      "6 1\n",
      "7 1\n",
      "8 1\n",
      "9 1\n",
      "10 1\n",
      "11 1\n",
      "12 1\n",
      "13 1\n",
      "14 1\n",
      "15 1\n",
      "16 1\n",
      "17 1\n",
      "18 1\n",
      "19 2\n",
      "20 3\n",
      "21 8\n",
      "22 17\n"
     ]
    }
   ],
   "source": [
    "lineages = df2['LINEAGE'].tolist()\n",
    "num_taxa = 20\n",
    "for i in range(len(lineages[0])):\n",
    "    to_add = []\n",
    "    for species in lineages:\n",
    "        to_add.append(species[i])\n",
    "    #end inner for\n",
    "    df2[\"Annotation\"] = to_add\n",
    "    a = df2[\"Annotation\"].nunique()\n",
    "    if a >= num_taxa: break\n",
    "    print(\"#\",i, \"=\", df2[\"Annotation\"].nunique())\n",
    "#end outer for\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4d9c133d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Annotation'>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.groupby('Annotation').Annotation.count().plot.pie(figsize=(10, 10), autopct='%1.1f%%', fontsize=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "518fc85b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([<AxesSubplot:ylabel='Annotation'>], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.groupby('Annotation').Annotation.count().plot.pie(figsize=(10, 10), autopct='%1.1f%%', fontsize=12, subplots=True, wedgeprops={\"edgecolor\":\"0\",'linewidth': 0.5})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "df85fe5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Saving  19 to: Carnivora.clade\n",
      "# Saving  1 to: Macroscelididae.clade\n",
      "# Saving  1 to: Trichosurus.clade\n",
      "# Saving  1 to: Didelphinae.clade\n",
      "# Saving  1 to: Folivora.clade\n",
      "# Saving  23 to: Glires.clade\n",
      "# Saving  1 to: Ornithorhynchus anatinus.clade\n",
      "# Saving  17 to: Chiroptera.clade\n",
      "# Saving  1 to: Tenrecinae.clade\n",
      "# Saving  25 to: Primates.clade\n",
      "# Saving  1 to: Scandentia.clade\n",
      "# Saving  1 to: Dermoptera.clade\n",
      "# Saving  1 to: Chrysochlorinae.clade\n",
      "# Saving  1 to: Trichechidae.clade\n",
      "# Saving  23 to: Artiodactyla.clade\n",
      "# Saving  3 to: Perissodactyla.clade\n",
      "# Saving  0 to: Elephantidae.clade\n",
      "# Saving  1 to: Phascolarctos.clade\n",
      "# Saving  1 to: Dasypodidae.clade\n",
      "# Saving  1 to: Sarcophilus.clade\n",
      "# Saving  3 to: Eulipotyphla.clade\n",
      "# Saving  2 to: Pholidota.clade\n",
      "# Saving  1 to: Orycteropodidae.clade\n",
      "# Saving  1 to: Vombatus.clade\n",
      "# Saving  1 to: Tachyglossus aculeatus.clade\n"
     ]
    }
   ],
   "source": [
    "for item in set(df2[\"Annotation\"].to_list()):\n",
    "    #print(item)\n",
    "    df3 = df2[df2[\"Annotation\"] == item]\n",
    "    #print(df3)\n",
    "    df3 = df3.dropna()\n",
    "    \n",
    "    #if len(df3[\"ACCESSION\"].to_list()) < 3: \n",
    "    #    continue\n",
    "    \n",
    "    #print(df3[\"LEAFNAME\"].to_list())\n",
    "    #df3.to_csv(item + \".csv\", index=True)\n",
    "    \n",
    "    print(\"# Saving\", str(len(df3[\"LEAFNAME\"].to_list())), \"to:\", item + \".clade\")\n",
    "    with open(item + \".clade\", \"w\") as fh:\n",
    "        fh.write(\"\\n\".join(df3[\"LEAFNAME\"].to_list()))\n",
    "        #for row in df3[\"LEAFNAME\"].to_list():\n",
    "        #    fh.write(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5287b897",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "154    Tachyglossus aculeatus\n",
       "Name: Annotation, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3[\"Annotation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c142578",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25566f1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# There may be errors above, some LEAFNAMEs do not populate. Check the search for subroutine to make sure it works."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
